algebird-core/src/main/scala/com/twitter/algebird/TDigest.scala

```diff
@@ -0,0 +1,263 @@
+/*
+Copyright 2015 Twitter, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package com.twitter.algebird
+
+import tdmap.TDigestMap
+
+/**
+ * A t-digest sketch of sampled numeric data, as described in:
+ * Computing Extremely Accurate Quantiles Using t-Digests,
+ * Ted Dunning and Otmar Ertl,
+ * https://github.com/tdunning/t-digest/blob/master/docs/t-digest-paper/histo.pdf
+ */
+case class TDigest(
+  delta: Double,
+  nclusters: Int,
+  clusters: TDigestMap) extends Serializable {
+
+  // re-cluster when number of clusters exceeds this threshold
+  private val R = (TDigest.K / delta).toInt
+
+  private case class Cluster(centroid: Double, mass: Double, massUB: Double)
+
+  /**
+   * Returns a new t-digest with value x included in its sketch; td + x is equivalent to
+   * td + (x, 1).
+   * @param x The numeric data value to include in the sketch
+   * @return the updated sketch
+   */
+  def +[N](x: N)(implicit num: Numeric[N]): TDigest = this.+((x, 1))
+
+  /**
+   * Returns a new t-digest with new pair (x, w) included in its sketch.
+   * @param xw A pair (x, w) where x is the numeric value and w is its weight
+   * @return the updated sketch
+   * @note This implements 'algorithm 1' from:
+   * Computing Extremely Accurate Quantiles Using t-Digests,
+   * Ted Dunning and Otmar Ertl,
+   * https://github.com/tdunning/t-digest/blob/master/docs/t-digest-paper/histo.pdf
+   */
+  def +[N1, N2](xw: (N1, N2))(implicit num1: Numeric[N1], num2: Numeric[N2]): TDigest = {
+    val s = this.update(xw)
+    if (s.nclusters <= R) s
+    else {
+      // too many clusters: attempt to compress it by re-clustering
+      val ds = scala.util.Random.shuffle(s.clusters.toVector)
```

Comment:
this is non-deterministic, but I wonder if we track a seed which we update on merges if we can make it deterministic and yet keep the property that in expectation it is randomized.